mkfs/rootdir.c

```diff
@@ -286,33 +397,108 @@ static int write_extent(struct btrfs_trans_handle *trans,
 		bytes_read += ret_read;
 	}
 
-	memset(buf + to_read, 0, padding);
+	if (bytes_read <= sectorsize)
+		do_comp = false;
+
+#if COMPRESSION_ZSTD
+	if (do_comp) {
+		comp_ret = compress_extent(btrfs_inode, sectorsize, buf,
+					   bytes_read, comp_buf);
+		if (comp_ret < 0)
+			return comp_ret;
+		else if (comp_ret == 0)
+			do_comp = false;
+
+		/* If we've just marked this inode nocompress, increase the
+		 * buffer size and read the rest of the extent.  */
+		if (btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_NOCOMPRESS) {
+			buf_size = SZ_1M;
```

Comment:
Can we just use 128K buf_size for all?

I think it's already a large enough buffer size. This buffer size change can be done in the previous commit.

Reply:
Nevermind 1MiB is much preferable to reduce fragmentation, so it's still better to use larger buffer size when possible.

My bad, it's not about read performance but fragmentation.